library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java

```diff
@@ -652,7 +652,42 @@ public boolean continueLoading(long positionUs) {
 
   @Override
   public void reevaluateBuffer(long positionUs) {
-    // Do nothing.
+    if (loader.isLoading() || isPendingReset()) {
+      return;
+    }
+
+    int currentQueueSize = mediaChunks.size();
+    int preferredQueueSize = chunkSource.getPreferredQueueSize(positionUs, readOnlyMediaChunks);
+    int chunksToDiscard = currentQueueSize - preferredQueueSize;
+    if (chunksToDiscard > 0) {
+      int removeRangeStart = currentQueueSize - chunksToDiscard;
+      long firstRemovedStartTimeUs = discardMediaChunks(removeRangeStart);
+      long endTimeUs = getLastMediaChunk().endTimeUs;
+      eventDispatcher.upstreamDiscarded(primarySampleQueueType, firstRemovedStartTimeUs, endTimeUs);
+    }
+
+  }
+
+
+  /**
+   * Discards, chunksToDiscard, HlsMediaChunks for the start of the list of mediaChunks.
+   *
+   *
+   * @param chunksToDiscard - Count of chunks to discard
+   * @return endTimeUs of last discarded chunk
+   */
+  private long discardMediaChunks(int chunksToDiscard) {
+    long lastDiscardEndTimeUs = mediaChunks.get(chunksToDiscard - 1).endTimeUs;
+
+    Log.d(TAG, "discardChunksToIndex() -  chunksToDiscard " + chunksToDiscard + " count " + mediaChunks.size());
+    for (HlsMediaChunk chunk : mediaChunks) {
+      Log.d(TAG, "chunk " + chunk.uid + " start/end: " + chunk.startTimeUs + "/" + chunk.endTimeUs + " format: "+chunk.trackFormat);
+    }
+
+    // TODO - need a way to discard just the samples owned by the discarded chunks
+    sampleQueues[primarySampleQueueIndex].discardToRead();
+    Util.removeRange(mediaChunks, 0, chunksToDiscard  - 1);
+    return lastDiscardEndTimeUs;
```

Comment:
This is just the first part of the implementation, because it is not yet possible to extend `ChunkSampleStream` this results in code copy from it.

It seems extending `ChunkSampleStream` is a better path for `HlsSampleStreamWrapper`.  Again, if that is the desired architecture I'm fine with taking that path.

Note the HlsSampleSteamWrapper's `mediaChunks` need to be discarded from index 0 to the index of the first 'good' format sample.  This seems not to be the way of `ChunkSampleStream#discardUpstreamMediaChunksFromIndex()`

Here's a snapshot of the logging output from above from an actual live stream that has up leveled.

````
08-19 16:27:06.749 21933 21967 D HlsSampleStreamWrapper: discardChunksToIndex() -  chunksToDiscard 1 count 4
08-19 16:27:06.750 21933 21967 D HlsSampleStreamWrapper: chunk 5 start/end: 1792000000/1796000000 format: Format(2, null, application/x-mpegURL, null, mp4a.40.2,avc1.64001F, 2100000, null, [960, 540, -1.0], [-1, -1])
08-19 16:27:06.751 21933 21967 D HlsSampleStreamWrapper: chunk 7 start/end: 1796000000/1800000000 format: Format(2, null, application/x-mpegURL, null, mp4a.40.2,avc1.64001F, 2100000, null, [960, 540, -1.0], [-1, -1])
08-19 16:27:06.751 21933 21967 D HlsSampleStreamWrapper: chunk 8 start/end: 1800000000/1804000000 format: Format(2, null, application/x-mpegURL, null, mp4a.40.2,avc1.64001F, 2100000, null, [960, 540, -1.0], [-1, -1])
08-19 16:27:06.752 21933 21967 D HlsSampleStreamWrapper: chunk 10 start/end: 1800000000/1804000000 format: Format(4, null, application/x-mpegURL, null, mp4a.40.2,avc1.640020, 8250000, null, [1280, 720, -1.0], [-1, -1])
````

So the samples and chunks with UID's 5,7,8 need to be discarded.

Reply:
To make that work, HlsMediaChunk needs to store a list of the first sample index for each track (similar to BaseMediaChunk). But I haven't checked in detail if and how we can do this.

Reply:
This has been done... See the latest checkin